src/analysis/processing/qgsalgorithmzonalstatisticsfeaturebased.cpp

```diff
@@ -0,0 +1,189 @@
+}
+
+QgsFields QgsZonalStatisticsFeatureBasedAlgorithm::outputFields( const QgsFields &inputFields ) const
+{
+  Q_UNUSED( inputFields )
+  return mOutputFields;
+}
+
+QgsProcessingFeatureSource::Flag QgsZonalStatisticsFeatureBasedAlgorithm::sourceFlags() const
+{
+  return QgsProcessingFeatureSource::FlagSkipGeometryValidityChecks;
+}
+
+bool QgsZonalStatisticsFeatureBasedAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  mPrefix = parameterAsString( parameters, QStringLiteral( "COLUMN_PREFIX" ), context );
+
+  const QList< int > stats = parameterAsEnums( parameters, QStringLiteral( "STATISTICS" ), context );
+  mStats = nullptr;
+  for ( int s : stats )
+  {
+    mStats |= STATS.at( s );
+  }
+
+  QgsRasterLayer *rasterLayer = parameterAsRasterLayer( parameters, QStringLiteral( "INPUT_RASTER" ), context );
+  if ( !rasterLayer )
+    throw QgsProcessingException( invalidRasterError( parameters, QStringLiteral( "INPUT_RASTER" ) ) );
+
+  mBand = parameterAsInt( parameters, QStringLiteral( "RASTER_BAND" ), context );
+  if ( mBand < 1 || mBand > rasterLayer->bandCount() )
+    throw QgsProcessingException( QObject::tr( "Invalid band number for BAND (%1): Valid values for input raster are 1 to %2" ).arg( mBand )
+                                  .arg( rasterLayer->bandCount() ) );
+
+  if ( !rasterLayer->dataProvider() )
+    throw QgsProcessingException( QObject::tr( "Invalid raster layer. Layer %1 is invalid." ).arg( rasterLayer->id() ) );
+
+  mRaster.reset( rasterLayer->dataProvider()->clone() );
+  mCrs = rasterLayer->crs();
+  mPixelSizeX = rasterLayer->rasterUnitsPerPixelX();
+  mPixelSizeY = rasterLayer->rasterUnitsPerPixelY();
+  QgsFeatureSource *source = parameterAsSource( parameters, inputParameterName(), context );
+
+  mOutputFields = source->fields();
+
+  for ( QgsZonalStatistics::Statistic stat : STATS )
+  {
+    if ( mStats & stat )
+    {
+      QgsField field = QgsField( mPrefix + QgsZonalStatistics::shortName( stat ), QVariant::Double, QStringLiteral( "double precision" ) );
+      if ( mOutputFields.names().contains( field.name() ) )
+      {
+        throw QgsProcessingException( QObject::tr( "Field %1 already exists" ).arg( field.name() ) );
+      }
+      mOutputFields.append( field );
+      mStatFieldsMapping.insert( stat, mOutputFields.size() - 1 );
+    }
+  }
+
+  return true;
+}
+
+QgsFeatureList QgsZonalStatisticsFeatureBasedAlgorithm::processFeature( const QgsFeature &feature, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  Q_UNUSED( context )
+  Q_UNUSED( feedback )
+  QgsAttributes attributes = feature.attributes();
+  attributes.resize( mOutputFields.size() );
+
+  QMap<QgsZonalStatistics::Statistic, QVariant> results = QgsZonalStatistics::calculateStatistics( mRaster.get(), feature.geometry(), mPixelSizeX, mPixelSizeY, mBand, mStats );
+  for ( const auto &result : results.toStdMap() )
```

Comment:
This looks inefficient -- why not iterate the QMap directly?

Reply:
Readability.
As soon as we bump to c++ 17 we'll be able to do that directly on the QMap (yay!)